components/util/src/datap/mod.rs

```diff
@@ -0,0 +1,186 @@
+// Data provider trait definitions
+
+pub mod decimal;
+
+mod helpers;
+
+use core::ops::Deref;
+use helpers::CloneableAny;
+use std::any::Any;
+use std::any::TypeId;
+use std::borrow::Borrow;
+use std::borrow::BorrowMut;
+use std::borrow::Cow;
+use std::error::Error;
+use std::fmt::{self, Debug, Display};
+use std::prelude::v1::*;
+
+/// A top-level collection of related data keys.
+#[derive(PartialEq, Copy, Clone, Debug)]
+pub enum Category {
+    Undefined = 0,
+    Decimal = 1,
+    PrivateUse = 4096,
+}
+
+impl Display for Category {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        Debug::fmt(self, f)
+    }
+}
+
+/// A specific key within a category.
+#[derive(PartialEq, Copy, Clone, Debug)]
+pub enum Key {
+    Undefined,
+    Decimal(decimal::Key),
+    PrivateUse(u32),
+}
+
+impl Display for Key {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        Debug::fmt(self, f)
+    }
+}
+
+/// A struct to request a certain hunk of data from a data provider.
+#[derive(Debug, PartialEq, Clone)]
+pub struct Request {
+    // TODO: Make this a Locale instead of a String
+    pub locale: String,
+    pub category: Category,
+    pub key: Key,
+
+    // For now, the request payload is a string. If a more expressive type is needed in the
+    // future, it should implement PartialEq and Clone. Consider using a concrete type, rather
+    // than a detour through Any (like in Response), which complicates the code.
+    pub payload: Option<Cow<'static, str>>,
```

Comment:
I'm not 100% sure about the future of this field.  Maybe it wants to be a TinyStr.  This also depends on the answer to #32, because ultimately we'll want the Request to be serializable down to a form that Fuchsia can use.